### DISCUSSION TITLE
Micro-frontends with SSI/ESI/Ajax: multiple Astro fragments from multiple Astro servers on one page

### DISCUSSION THREAD

**Discussion by philipschm1tt** [withastro/roadmap#713] [Proposal]
### Body

# Summary

A classic micro-frontend approach for server-rendered pages uses Server-Side Includes (SSI), Edge-Side Includes (ESI), and/or Ajax to compose a page from multiple different servers. Astro does not support this well: multiple instances could clash on the client side because of global state/elements/events. I propose to find a way to avoid these clashes – maybe some configurable names for global things would do the trick.

# Background & Motivation

Micro-frontends are gaining traction.
The latest InfoQ trend report shows them in the early adopters column (see https://www.infoq.com/articles/architecture-trends-2023/ ) and there are many well-known large companies that have adopted micro-frontends.

Micro-frontends are a great fit for current architecture trends in general.
When organizing teams "for fast flow" – aligned with value streams (see Team Topologies) and using Domain-Driven Design, the idea is that teams can work autonomously in their domains. Ideally, a team could make changes to the backend and frontend within their domain without hand-offs to other teams.

In the backend, we have long moved towards independent services.
It would be unusual to see 20 teams working on one (backend) monolith. 
It would be inconsistent if we stick with monolithic frontends.

Composing multiple fragments from different applications on a single page is relatively simple with MPA-style frontends. Where SPAs require complex solutions with application shells like single-spa or piral, some techniques for the "MPA" world are simple, battle-tested and widely available: classic approaches like Server-Side Includes or Edge-Side Includes.

The problem: Astro is not designed to support two or more Astro fragments on one page. It uses global pieces like the astro-island custom element and global events. The dev and preview modes rely on fix paths like `/src` that cannot be changed.

While Astro is not designed for that use case, there is some interest in it.
Some comments on https://github.com/withastro/roadmap/discussions/266 refer to micro-frontends.

# Goals

- Enable two or more Astro fragments to work on the same page without conflicts/clashes between them
- ... without using global state/element/events.
- ... even when the applications use different versions of Astro (you don't want to force multiple teams to upgrade their versions in lock-step)
- Enable dev/preview mode for pages that contain multiple Astro fragments

Note: the dev/preview mode has lower priority; with micro-frontends it is a good pattern to test fragments in isolation.

# Questions

Might there be a relatively lightweight way to "retro-fit" Astro with that capability?
What aspects of Astro use global state/elements/events? Might it be possible to introduce hashes in the names of these state/elements/events so they don't interfere with another instance? 

# Relates to 

- https://github.com/withastro/roadmap/discussions/266
- https://github.com/withastro/roadmap/issues/697

**Comment by lilnasy** [comment]
For some people, astro has already been the solution to what they call micro frontends.

https://github.com/sasoria/astro-microfrontends-ssr
https://github.com/ec965/micro-frontends-astro

It would help if you could share the specifics of your use case because it's a fairly new term and definitions vary.

**Reply by philipschm1tt** [reply]
I renamed the title from "Support micro-frontends: multiple Astro applications on one page" to "Support Astro micro-frontends: multiple Astro applications on one page" to be a bit more clear.

I don't think the definition for micro-frontends vary much – it's just that there are different approaches to micro-frontends. I will add some more context here.

### Context on micro-frontends

First some more general context on micro-frontends and micro-frontends with Astro; some of it inspired by the book Micro-Frontends in Action (that I prefer over Building Micro Frontends).

**Micro-frontends**: a single website is in some way composed from pieces that are independently developed/built pieces.

There are a number of approaches that qualify as micro-frontends. Some of them are framework-agnostic and can be used in environments that include Astro. A non-exhaustive list of examples:
- **Links**: you can build a website with two web applications that you connect only via links; maybe with a reverse proxy to route to the right application. 
- **Iframes**: you can use classic iframes to get pieces from separate web applications on the same web page. 
- **Web components**: you can use web components to to get a fragment built by team A onto a page built by team B.
- **Ajax**: you can include a fragment from another web application by requesting via client-side JavaScript the HTML, script tags, and stylesheet links for that fragment and adding them to the markup.
- **Server-side integration**: you can use technologies like Server-Side Includes (SSI) to include fragments from different web applications on the same page; that might happen in an nginx; the fully server-side rendered page would be sent to the clients.
- **App shell**: you can use a client-side application shell like single-spa that performs client-side routing and can compose pages from multiple frontends.

So there are different approaches to routing (links, client-side, server-side) and to composing fragments from different frontends on the same page (iframe, ajax, client-side, server-side).

### Micro-frontends with Astro

A number of techniques can be easily combined with Astro. For example:
- You can link two Astro applications
- You can include web components on an Astro page
- You can use iframes within an Astro application or to compose multiple Astro applications

The central question is, what technique(s) fit best for your context.
In the Astro world, that context typically involves server-rendered pages with little client-side JavaScript (some islands of interactivity).

#### What about routing?

The routing part of micro-frontends is easy. You can use links to connect different Astro applications and you can use a reverse proxy to route to different Astro applications depending on the path.

#### What about composition?

You may want to compose fragments from different frontends on the same page.
In a simple case, you may have a common header and footer that you want to have on all pages.
In a more advanced case, you may want to include some product recommendations from a recommendations team on different pages on an ecommerce site (like on the home page, the product details page, and the cart page).

- You probably don't want to use iframes on modern B2C sites (due to annoyance with responsive design and accessibility).
- You could use Web Components for fragments on a page, but are the client-side rendered components a good fit for the Astro world? You probably don't want you header jumping around on page changes.
- Server-side integration would be a good fit in the Astro world: if you could splice together a server-rendered page from multiple frontends that contribute their fragments and use little JavaScript client-side, that seems like a good fit to me.
- Ajax could be a good fit for some lazy-loaded composition.

The problem: Astro does not support multiple Astro applications on one page.

So using server-side integration or Ajax, the teams could not use two Astro applications on one page.
Maybe they would have to go back to the old ways of doing "MPAs" with like some Java templating and a separate JavaScript bundle.

So are server-side integration and Ajax the wrong mechanisms for micro-frontends with Astro even though they focus on server-rendered pages/fragments with little client-side JavaScript?

**Reply by philipschm1tt** [reply]
### Specific use case?

I'm coming from the context of a large ecommerce website with more than a dozen teams working on the web frontend. We use Domain-Driven Design and teams that are as much as possible aligned with value-streams: a recommendation team could ideally provide both the backend as well as the UI for their features so that they can work autonomously on their domain. The same type of recommendations (like a slider with a number of products) might be included on a number of pages that belong to different teams. 

For this ecommerce website, we decided that an approach with server-rendered pages – ideally with limited amounts of client-side JavaScript. And we are not the only ones: Amazon, Ebay, Zalando, ... have shown that this direction leads to favorable business outcomes. And that's where Astro is particularly interesting to us!

At our size and number of teams working on the web frontend, micro-frontends are appealing to us.
With many teams, a frontend monolith can be cumbersome for collaboration/coordination/releases and it can make it difficult to adopt new technologies or upgrade frameworks, since you might have to upgrade everything in one go.

We have explored Server-Side Includes as a classic, "battle-tested" mechanism to compose fragments on a page.
And so far we had been hoping that some of these fragments could use Astro.

Using a module federation approach like in https://github.com/sasoria/astro-microfrontends-ssr would technically allow us to compose a page from multiple frontends, but my impression of this approach is that it does not quite match the Astro way. Wouldn't you forego the idea of islands and minimal JavaScript and get larger amounts of client-side JavaScript and hydration?

**Reply by lilnasy** [reply]
Thanks for the detailed response! It really helps move the conversation.

<details>
<summary><blockquote>Links, iframes, web components: get pieces from separate web applications on the same web page</blockquote></summary>

If I'm understanding correctly (see example below), this is well-supported
```astro
---
import { VueHeader } from "@company/team-header-components"
import { MainContent } from "../components/homepage.astro"
import ChatBoxOverlay from "@company/chatbox/overlay.tsx"
import { SolidFooter } from "@company/team-footer-components"
---
<VueHeader client:load />
<MainContent />
<ChatBoxOverlay for="onboarding" client:idle />
<SolidFooter />
```

</details>

<details>
<summary><blockquote>Ajax</blockquote></summary>

You can create a page that you only intend to access via fetch calls instead of direct browser navigation. It's possible already, and better ergonomics are being considered in https://github.com/withastro/roadmap/issues/697.

</details>

<details>
<summary><blockquote>Server-side integration</blockquote></summary>
Not sure I understand this one. If you can compose markdown, MDX, SVG, plain HTML, framework, and web components into a single page, maybe you wouldn't need SSI in the first place. And this composability is the foundation of Astro.
</details>

<details>
<summary><blockquote>The problem: Astro does not support multiple Astro applications on one page.</blockquote></summary>
I think this could be reiterated. I don't see it as a problem, but a preconceived solution.
</details>

<details>
<summary><blockquote>With many teams, a frontend monolith can be cumbersome</blockquote></summary>

Keeping the codebases separate with Astro is trivial.

But if you must keep the servers and deployment lifecycles separate, it becomes difficult.

One idea is to give each team a subpath: `example.com/app/*` routes to team A's server, `example.com/blog/*` routes to team B's server. Whether it's a good idea to split that way, would depend on how the teams are split.

Another idea is to let each team host internally accessible fragment servers, and the publicly accessible main server could make fetch calls for the fragments. Based on what you've said so far, this seems closer to your organization structure. You may want to chime in on https://github.com/withastro/roadmap/issues/697.

</details>

<details>
<summary><blockquote>my impression of this approach is that it does not quite match the Astro way</blockquote></summary>
It very much does ;)
</details>

<details>
<summary><blockquote>Wouldn't you forego the idea of islands and minimal JavaScript and get larger amounts of client-side JavaScript and hydration?</blockquote></summary>
The javascript remains opt-in per component use, so I wouldn't say so.</details>

**Reply by philipschm1tt** [reply]
<details>
<summary>

> If I'm understanding correctly (see example below), this is well-supported
</summary>

Right. I mentioned that above: "A number of techniques can be easily combined with Astro. For example:"
It would be tragic if a web application framework broke web standards including links :)
</details>

<details>
<summary>

> You can create a page that you only intend to access via fetch calls instead of direct browser navigation. It's possible already, and better ergonomics are being considered in https://github.com/withastro/roadmap/issues/697.
</summary>

With Ajax, I already have some doubts. I know the goal to better support partials.
But what if you have an Astro application and then want to use Ajax to include a partial from another Astro application?
With Ajax, you ideally shouldn't have to care what framework it was built with: you just want an HTML fragment, a script tag, and a stylesheet, to embed a fragment and make it interactive. Then we would run into the problem that there are two Astros on one page that share some global stuff.
</details>

<details>
<summary>

> Not sure I understand this one. If you can compose markdown, MDX, SVG, plain HTML, framework, and web components into a single page, maybe you wouldn't need SSI in the first place. And this composability is the foundation of Astro.
</summary>

Yes, you can compose a page using multiple technologies and when that is sufficient, you can save yourself the trouble of using micro-frontends. But at what point do you have two many teams working on the same application and how do you deal with it?

My goal is not to throw as many technologies as possible on one page.

My current goal: I want to divide an ecommerce site into a few frontend applications: one for the checkout, one for searching and browsing the product catalog, and one for your account. In each of these applications, you might have a handful of teams working on them. But you would also need to use the same header and footer across these few applications. So may the header and footer could be separate micro-frontends. That means that on any given page, I would have the main content of the page from only one frontend but then also use the common header and footer from separate micro-frontends. And SSI could include header/footer on the page.
The technologies might be pretty homogenous: all teams use the same component framework; and header/footer and the part in-between might all use Astro; except for the checkout that uses some SPA technology for better payment integration.

My ideal goal: ideally, I would use a bit more composition. In addition to header and footer, there may be some fragments in the content of the page that belong to a different domain/team. That domain/team could have their own frontend. Maybe using Astro, maybe using some Java templating.
</details>

<details>
<summary>

> I think this could be reiterated. I don't see it as a problem, but a preconceived solution.
</summary>

I would like to have 20+ teams working on a website pursuits Astro's goals: 1) modern DevEx without building separate server and JavaScript bundle, 2) server-rendered pages 3) little JavaScript and mainly for the "islands of interactivity".
But with that many teams, working on a monolith becomes an issue. We've largely stopped doing that in the backend; the frontend is not that different. I believe no one frontend framework can solve that: at some point you get benefits from splitting it a bit.
</details>

**Reply by philipschm1tt** [reply]
> One idea is to give each team a subpath: example.com/app/* routes to team A's server, example.com/blog/* routes to team B's server. Whether it's a good idea to split that way, would depend on how the teams are split.
> Another idea is to let each team host internally accessible fragment servers, and the publicly accessible main server could make fetch calls for the fragments. Based on what you've said so far, this seems closer to your organization structure. You may want to chime in on https://github.com/withastro/roadmap/issues/697.

Yes, thanks, these are **exactly** the things I'm interested in.

The approach with the subpaths is pretty much my current goal. I would split the application into like three pieces with a handful of teams working on each piece. But what about the header and footer? How do I share those across the three pieces? In a way that does not cause JavaScript clashes when the pieces don't upgrade their Astro version at exactly the same time. And in a way that does not impact the loading behavior (layout should not jump) and performance should stay high.

About the fragment servers: that would be great: I would love to load header/footer from a fragment server into an Astro application. But that brings me to my Ajax doubts from above: if I load a fragment from a different Astro instance, I could run into clashes between the two Astros.

**Reply by philipschm1tt** [reply]
<details>
<summary>
Using a module federation approach like in https://github.com/sasoria/astro-microfrontends-ssr would technically allow us to compose a page from multiple frontends, but my impression of this approach is that it does not quite match the Astro way. Wouldn't you forego the idea of islands and minimal JavaScript and get larger amounts of client-side JavaScript and hydration?

> It very much does ;)
> The javascript remains opt-in per component use, so I wouldn't say so.
</summary>

I'm not fully opposed to the approach. I think it could be interesting to get my header and footer onto the page (if the SSR works).

But generally: I believe that goes into the client-side direction where some (potentially large and complex) fragments on a page are are pure React/Angular/Vue (in my case Vue) applications even when only small parts of the fragment are interactive. You can probably still treat the whole fragment as an island and decide whether to hydrate it, but the code and hydration will cover the whole fragment and not just the interactive islands within the fragment.
The micro-frontend fragments are not scoped according to interactivity but by the "domain" of the fragments. 

For our micro-frontend journey I read some books on the topic: Micro-Frontends in Action by Michael Geers and Building Micro Frontends by Luca Mezzalira. Both feature a chapter to help you decide on the architecture / technical approach for micro-frontends. In short: approaches with SSI and/or Ajax are a good, recommended approach for the composition on "MPA" style pages.

In less short: 

For linked pages (one or multiple "MPAs"), Michael Geers decision tree shows "Server-side integration and/or Ajax" as the recommended approach for composition (one one page).
![micro-frontends_architecture_mfia](https://github.com/withastro/roadmap/assets/8048758/cc638c35-0db9-47d1-999f-f169b072b45f)

For websites that use composition of micro-frontends on one page, Luca Mezzalira shows a path from server-side composition to server-side routing, to server-side includes (SSI) and server-side rendering.
![micro-frontends_architecture_bmf](https://github.com/withastro/roadmap/assets/8048758/93ef4255-f93a-4b31-81c4-ab109f22660e)
</details>

**Reply by sasoria** [reply]
To clarify further, the ES-Module approach that's mentioned here would be completely SSR-ed with the possibility to opt-in to client side hydration on a per-component basis with something like SolidJS or Preact's [hydrate()](https://preactjs.com/guide/v10/api-reference/#hydrate) function. It is however true that an Astro-in-Astro solution would make this easier for other frameworks.

**Reply by lilnasy** [reply]
> But what about the header and footer?

Shared components can go in an npm package imported by each subpath-project.

**Comment by sasoria** [comment]
For reference, here's an example with client-side composition.

https://github.com/sasoria/astro-microfrontends

**Reply by philipschm1tt** [reply]
Thanks – I had already seen your prototype as well as the two that @lilnasy linked (https://github.com/sasoria/astro-microfrontends-ssr and https://github.com/ec965/micro-frontends-astro) in the Discord. It's good to have them here, too, so that people looking for micro-frontend options get a good overview.

The thing is that the micro-frontend approaches that I associate the most with Astro's goals and concepts – server-side integration and Ajax – don't appear to allow multiple Astro applications. That would be valuable in my context.

**Comment by philipschm1tt** [comment]
@ematipico, I take it (based on the thumbs-down emoji) you don't like the idea of supporting server-side integration or Ajax integration of multiple Astro frontends? Why is that?

**Reply by ematipico** [reply]
It's not that I don't like it. As @lukastaegert mentioned, the micro-frontend architecture is the byproduct of teams and organisations, and that's a fact.

The micro-frontend is essentially a way to have multiple **frontend** libraries in the same application/website. We can go into specifics in a separate thread, but I believe we agree that everything boils down to that. 

I put the thumb-down emoji mainly because the proposal needs to change to fit the Astro framework, and Astro **is not** a _frontend_ framework, Astro is a **backend** framework first and foremost. While it's true that we have a  runtime and it's possible to ship JS code to the client, we only provide a few features that frontend libraries/frameworks do, and if we do, our approach uses the backend.

For example, nowadays, Astro doesn't provide a way to share state clientside, because Astro let you choose what client-side library you want.

So I believe we should make a proposal tailored to Astro.

>  The dev and preview modes rely on fix paths like /src that cannot be changed.

I believe it's possible: https://docs.astro.build/en/reference/configuration-reference/#srcdir

**Reply by philipschm1tt** [reply]
<details>
<summary>

> essentially a way to have multiple frontend libraries in the same application/website.
</summary>

Depending on what you mean, we can possibly agree :)
I just want to clarify that we have one standard component framework across teams (Vue.js) and we are not adopting micro-frontends to change that. We mainly want to stop working in a monolith with more than a dozen teams and make it easy to do migrations and technology changes incrementally – not having to upgrade a huge monolith all at once.

If all my teams were using Astro and Vue.js with a small number of Astro instances, I would be happy.
</details>

> I put the thumb-down emoji mainly because the proposal needs to change to fit the Astro framework, and Astro is not a frontend framework, Astro is a backend framework first and foremost. 

Maybe I can reword some aspects of the post to frame Astro less as a frontend.
I'll have a look.

But I don't understand how you would tailor the solution to the Astro framework.
To me, it seems like Astro does not play well with some common web technologies and techniques (SSI, ESI, Ajax) and it might not be hard to achieve that compatibility (by being able to rename a few global client-side things).

In my case, Astro is not competing frameworks like React, but with frameworks like Spring / Spring MVC.
When our ecommerce teams build a micro-frontend fragment in Java with Spring, there is no problem at all when using SSI to include other fragments written in Spring or whatever other technology. 
But then you have to deal with the "classic MPA" way of building frontends with markup in Java land and a separately built JavaScript bundle. I love how much Astro is improving the developer experience for building these server-rendered pages with little JS. So I would _love_ to adopt Astro in my ecommerce context!

In the ideal world, when a page includes a fragment from another server via SSI or Ajax, the team should not have to know what technology the server is using to deliver HTML, CSS, and JS. The fragment that is being included should avoid using global state/elements/events so it does not clash with other fragments.

**Reply by philipschm1tt** [reply]
I reworded parts of the proposal including the title and summary.

**Comment by maddsua** [comment]
Genuinely I don't get these modern trends. First we create serverless and micro services, then, a few years later, we complain about the complexity of our systems and the skyrocketing costs of running it on scale. Now the same thing is happening to the frontend.

**Reply by philipschm1tt** [reply]
The "micro" can be misleading in both frontend and backend.

You are right that the industry has learned that splitting any system into many tiny services/functions is not a great idea. In my context, we are aiming for roughly one-service-per-team or small-ish modular services that a few teams work on. My goal for the frontend is the same.

Just like you can go too small, you can also go to big.
How many teams would you put to work on one monolith? At what point would you split it up at least a bit?

**Reply by maddsua** [reply]
You see, that one of the problems with buzzwords. Like, is having subdomains with different apps deployed on them counts as a micro frontend? Or if redirects are set up in such a way that some sections of a website are in fact loaded from different domains and are powered by different builds/apps?

**Reply by maddsua** [reply]
To be completely honest, from the messages I've read and the examples of these micro frontends, to me it all boils down to code organization.
Let's look at https://github.com/sasoria/astro-microfrontends-ssr, aren't they just doing Islands here?

**Reply by philipschm1tt** [reply]
> Like, is having subdomains with different apps deployed on them counts as a micro frontend? Or if redirects are set up in such a way that some sections of a website are in fact loaded from different domains and are powered by different builds/apps?

These areas are where some people see the term "micro-frontend" differently.
I agree with the view from the book Micro-Frontends in Action: they are micro-frontends.
The key is that the frontend is made up of different deployment units or builds.

**Comment by lukastaegert** [comment]
For me, the main reason to have micro-frontends despite the obvious additional overhead with regard to deployed JavaScript size and the additional infrastructure cost is team autonomy.
This does not make sense for smaller projects, but at some point when people are split across different departments with disjoint areas of responsibility having

* separate repositories with
* separate build and deployment processes

can be a huge time-saver. The teams can make their own technology decisions, code style and organization decisions, maintain their own pipelines and processes and perform independent releases. It allows teams to make important decisions independently, greatly improving time-to-market.

I do not think we need a discussion whether this is a good idea. For some projects it is the right approach while for most projects it certainly isn't. It is a question of measuring how much the inter-team friction costs you.

The obvious downside of such an approach is that frameworks and dependencies are duplicated, and while there are ways of resolving that duplication e.g. via module federation techniques, this again creates additional overhead and threatens to eat up the independence advantage we already gained.

This is where Astro comes in as it has many features that both reduce deployed JavaScript as well as it allows features to be easily lazy-loaded on demand. So I would see it as a match made in heaven if this was possible, and I am quite sure that it would drive adoption for large business websites.

However, there were some issues we identified, which is why this issue was brought forward to create some interest on your side. Maybe solving them is as simple as allowing some names to be configured so that astro applications do not clash?

**Comment by sasoria** [comment]
I think this is an interesting proposal, but it might be useful to break it down to two pieces.

1) Is it possible to have an Astro-in-Astro app with the work that is being done on Astro partials?
2) Is it possible to have an Astro-in-Astro app with SSI/ESI/Ajax?

Maybe this would make it easier for the Astro maintainers to decide how to evaluate this proposal. This also brings up other questions: if (1) is possible, then why would you want to adopt (2)? Is it because ESI supports fallbacks and caching mechanisms are easy to implement? Would (2) require anything from Astro?

**Reply by philipschm1tt** [reply]
There is now some progress on the feature to support partials with this RFC: https://github.com/withastro/roadmap/pull/721

The proposal does not aim to solve Astro-in-Astro apps. 
The fragments would be served without non-inline scripts and without scoped styles. 
So it would not break anything if you served that fragment from another Astro but if the styles and scripts need to be on the page already, then they should come from the same Astro server as the fragment.

**Comment by philipschm1tt** [comment]
@ematipico @lilnasy, since you probably have a good overview of Astro: are you aware of more things that happen client-side in a way that two Astro instances could clash?

We know of the...
- astro-island custom element
- the astro:only even on load
But maybe there is more?

You might also have a better feel for how problematic those things could be.

The custom element might not be a big deal (unless a future Astro version changes it). The element would only be registered once and if the Astro fragments on the page would use the same Astro (major?) version, it probably just works.

The astro:only event may be more interesting, since you might dispatch it twice (from both Astro pieces).

If we had an idea, what things could cause issues, then we could also judge how simple it might be to work around the issues. If in the end it's just a small number of element/event names that could get something like a configurable suffix, this may be some "low-hanging fruit", and require little design effort. An interested party might even be able to contribute to make it happen.

**Reply by lilnasy** [reply]
Their existence means you can't call your own custom element `astro-island`, or dispatch events with the topic `astro:hydrate` or `astro:page-load`.

I don't see how either of those could interfere in another way.

**Reply by philipschm1tt** [reply]
<details>
<summary>

Their existence means you can't call your own custom element astro-island
</summary>

I'm not super worried about astro-island, but the critical scenario would be:
- Two Astro servers contribute fragments (that contain islands) to the same page
- Astro releases a new version that makes a significant change to `astro-island`
- One team is faster at upgrading to Astro – the teams can't all upgrade at the same time
- Now you have a page where one version of `astro-island` wins over the other one and you don't know how the two Astro fragments affect each other

I think for my current use case with header and footer as Astro fragments, I could mitigate that risk.
We could ensure that header and footer always use the same version as the main content on the page – even if we have to keep two versions of the header/footer for a little while.
</details>

<details>
<summary>

or dispatch events with the topic astro:hydrate or astro:page-load.
</summary>

The events seem a bit scary to me because I don't know their effects and side effects.

The most basic scenario:
- Two Astro servers contribute fragments (that contain islands) to the same page
- The servers may even use the same Astro versions

Now what happens during page load?
Might the events be triggered twice (once from each fragment) and cause problems?
Might the event from the first fragment come at a time where the second fragment is not in the right state and cause problems?

From the outside I have no clue what to expect here and what risks this might pose.
And a page wish mixed versions (one team upgrades Astro quicker) might also complicate things.

But if Astro contributors and maintainers who have a good overview of the behaviors could tell me what to expect, then that may be an acceptable risk. 

The risks and the required effort to affects the "priority" of this proposal.
I have some hopes that the effort to make the names of global elements configurable or application-name-specific might be relatively small.
If the risk is considerable and it helps large ecommerce sites adopt Astro in a big way, the "priority" goes up.
If there is not much risk, then maybe the proposal is not valuable and we could use Astro in our fragments.
</details>

**Reply by lilnasy** [reply]
You're right, you will have to sync the Astro version bump when prop serialization changes. I don't think there's a way around it.

The events only come into the play when you use view transitions. They could be problematic for your use case, but they are not there by default.

Edit: ...when you use view transitions in SPA mode.

The other events in the codebase are internal. `astro:hydrate` and `astro:unmount` – both contained within the associated `astro-island`. They do not bubble. They are not global.

**Comment by y-nk** [comment]
Could you provide a (mono)repo with 2 or more astro apps and a meta one making SSIs or at least a schematic of it and _show/describe_ the pain points in comments?

**Reply by philipschm1tt** [reply]
Yes, probably (in a week or two). I believe a colleague has a PoC with two Astros that may be suitable.
I don't think you'd see much but it might still be helpful to have a PoC for this discussion.

Based on the thread above I don't think there would be any immediate issues but as soon as a new Astro version would change the astro-island custom element and the fragments would upgrade at different times, something would break: one of the fragments would register their astro-island custom element and it would act on the whole page. In a larger micro-frontend context it would be a problem to upgrade all Astros as the same time: multiple teams owning multiple fragments on various pages – you want them to be able to upgrade their dependencies autonomously at their own pace.

**Reply by Lonli-Lokli** [reply]
Can you also add some react or vue app as the third one?

**Reply by reneleonhardt** [reply]
@philipschm1tt Did you have the chance for a PoC? 🙂
